### PR TITLE
fix PluginDirectory

### PR DESCRIPTION
**Pull request by FernetMenta**
@MilhouseVH this should fix http://trac.kodi.tv/ticket/16718

### CODE REVIEW

**Comment by FernetMenta** [comment]
@MartijnKaijser this pr has a slight rework for https://github.com/xbmc/xbmc/pull/9748
maybe you are lucky and it fixes the Android issue.

**Comment by MilhouseVH** [comment]
http://trac.kodi.tv/ticket/16718 appears to be fixed by this PR - many thanks!

**Comment by FernetMenta** [comment]
jenkins build this please

File: xbmc/windows/GUIMediaWindow.cpp [hunk 1]
```diff
@@ -1486,7 +1492,21 @@ void CGUIMediaWindow::OnInitWindow()
 
   // the start directory may change during Refresh
   bool updateStartDirectory = URIUtils::PathEquals(m_vecItems->GetPath(), m_startDirectory, true);
-  Refresh();
+
+  // we have python scripts hooked in everywhere :(
+  // those scripts may open windows and we can't open a window
+  // while opening this one.
+  // for pligin sources delay call to Refresh
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: typo here. pligin -> plugin

**Comment by MartijnKaijser** [comment]
No change

**Comment by MilhouseVH** [comment]
Although the Search YouTube hang appeared to be resolved, it's possible this PR is causing other odd behaviour when starting addons (or it's another hangover from the original problem, #9664). 

With PR9778, sometimes the user interface for the addon will "flash" (appear briefly then disappear), other times the addon will simply show a blank background (but the GUI hasn't hung - "Back" will eventually return to the Estuary Home menu).

I can reproduce this with latest master + PR9778 and YouTube addon in Ubuntu when running Estuary Home > Addons > YouTube.

Another addon showing similar problems is the Amazon Prime Instant addon.

In fact when you run the Amazon Prime Instant addon you see only a blank background. Click "Back" and you see a constantly spinning "Busy" icon. Click "Back" again and you're returned to the Estuary Home menu. Now click on Search > Search YouTube and you'll be shown the Amazon Prime Instant addon screen overlaid with a "Busy" dialog...

File: xbmc/windows/GUIMediaWindow.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/windows/GUIMediaWindow.cpp]: thanks, corrected

**Comment by FernetMenta** [comment]
@MilhouseVH updated. could you test again please

**Comment by MilhouseVH** [comment]
Sorry for the delay, updates look good @fernetmenta have tested all the add-ons on my RPi3 system and all are working as expected. Thanks again!

**Comment by FernetMenta** [comment]
thanks for feedback. there is one minor fix I am going to add to this PR before merge. currently you have to cancel twice when the keyboard dialog is displayed.

**Comment by MilhouseVH** [comment]
> currently you have to cancel twice when the keyboard dialog is displayed.

For example the YouTube addon? Yes I see that, first "Back" (or Cancel) shows a busy dialog. Strangely I don't get this issue with Global Search, so I assumed it was specific to the YouTube addon (ie. searching for `""`).

**Comment by FernetMenta** [comment]
> For example the YouTube addon? Yes I see that, first "Back" (or Cancel) shows a busy dialog

Exactly this. The old code did observe script termination. I will add this back but slightly different.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by HitcherUK** [comment]
Same problem here with script filled dynamic lists triggering the busy dialog.

@ronie @phil can you confirm the same?

**Comment by FernetMenta** [comment]
scripts are not supposed to hijack the main thread for a long time. if you see the busy dialog, you are doing something wrong

**Comment by Ignoble61** [comment]
I don't believe this is a case of plugins 'hijacking' the main thread, but rather when a list is filled directly from a plugin (`<content>plugin://plugin.name/</content>`) this now triggers the busy dialog where it didn't previously. It causes major usability issues with skins which directly fill lists from plugins as widgets, and definite usability issues with all skins which use @phil65's autocompletion plugin with the virtual keyboard, including Estuary.

If plugins are now returning their contents on the main thread, then this would appear to be a change of behaviour leading to a regression in terms or usability, and not the result of any change in plugin behaviour - and would affect all plugins I have tested.

**Comment by ronie** [comment]
> @ronie @phil can you confirm the same?

yup, can confirm.

perhaps we can grant plugins a little more time to provide a listing before popping up the busy dialog?
if i read the changes correctly, it used to be 1500 msecs before a progressbar was shown.

**Comment by phil65** [comment]
I think for dynamic content included via `<content>plugin://plugin.name/</content>` there never was a progress/busydialog. Those dialogs should only appear if the list which is loading content is our main media list (Video/Audio plug-ins for example)

**Comment by FernetMenta** [comment]
1)
the old design was wrong and caused issues. that was the reason why I fixed it. 

2)
GetDirectory must not silently hijack the main thread. 

3)
if the current behaviour does not fit your requirements, write those down in the dev section of the forum

**Comment by phil65** [comment]
just fyi, i had a talk with @FernetMenta today and explained the "problem" in more detail. He is aware of all the details now and will come up with a sane fix for this issue soon I think. :)

**Pull request merged**